overlord/devicestate/systems.go

```diff
@@ -176,3 +186,195 @@ func seededSystemFromModeenv() (*seededSystem, error) {
 	}
 	return seededSys, nil
 }
+
+func createSystemForModelFromValidatedSnaps(getInfo func(name string) (*snap.Info, bool, error), db asserts.RODatabase, label string, model *asserts.Model) (newFiles []string, dir string, err error) {
+	if isUC20 := model.Grade() != asserts.ModelGradeUnset; !isUC20 {
+		return nil, "", fmt.Errorf("cannot create a system for non UC20 model")
+	}
+
+	logger.Noticef("creating recovery system with label %q for %q", label, model.Model())
+
+	// TODO: should that path provided by boot package instead?
+	recoverySystemDirInRootDir := filepath.Join("/systems", label)
+	assertedSnapsDir := filepath.Join(boot.InitramfsUbuntuSeedDir, "snaps")
+	recoverySystemDir := filepath.Join(boot.InitramfsUbuntuSeedDir, recoverySystemDirInRootDir)
+
+	wOpts := &seedwriter.Options{
+		// RW mount of ubuntu-seed
+		SeedDir: boot.InitramfsUbuntuSeedDir,
+		Label:   label,
+	}
+	w, err := seedwriter.New(model, wOpts)
+	if err != nil {
+		return nil, "", err
+	}
+
+	optsSnaps := make([]*seedwriter.OptionsSnap, 0, len(model.RequiredWithEssentialSnaps()))
+
+	// collect all snaps that are present
+	modelSnaps := make(map[string]*snap.Info)
+	for _, sn := range model.EssentialSnaps() {
```

Comment:
This should probably use RequiredWithEssentialSnaps(). However, in the testing I don't see either of calls return snapd as a model snap. Although when testing with a device, I could snapd being pulled in when I tested in a VM with ubuntu-core-20-amd64-dangerous model, as snapd is listed explicitly there.

Reply:
as we discussed elsewhere this is fine as is but the code needs to deal with implicit snapd

Reply:
Pushed a patch that deals with implicit snapd.